teps/0075-object-param-and-result-types.md

```diff
@@ -0,0 +1,431 @@
+     * Potential groupings for the git-clone task stand out as well, for example the proxy configuration
+     * On that note, since the git-rebase task is using git and accessing remote repos, it probably needs the same
+       proxy configuration, so what they probably both need is some kind of interface for what values to provide when
+       accessing a remote git repo
+   Other examples include [okra-deploy](https://github.com/tektoncd/catalog/tree/main/task/orka-deploy/0.1) which is using
+   param name prefixes to group related params (e.g. `ssh-`, `okra-vm`, `okra-token`).
+
+## Requirements
+
+* Must be possible to programmatically determine the structure of the object
+* Must be possible for a result object to be empty
+* Must be possible to use the object results of one Task in a Pipeline as the
+  param of another Task in the pipeline which has a object param
+* Must be possible to use one specific value in a object result of one Task
+  in a Pipeline as the param of another Task in the pipeline which has a string param
+  * If there is no value for the specified key, the Pipeline execution should fail
+    (or we may be able to use [TEP-0048](https://github.com/tektoncd/community/pull/240)
+    to specify a default value to use in that case instead)
+
+## Proposal
+
+* We would add support for object types for results and params, in addition to the existing string and array support
+  * Initially we would only support string keys, eventually we can expand this to all values (string, array, object)
+    (note that this is the case for arrays as well, we don't yet support arrays of arrays)
+* We would use [json object schema syntax](https://github.com/tektoncd/pipeline/issues/1393#issuecomment-558833526)
+  to express the object structure
+* To support object results, we would support writing results to a `.json` file, as described in
+  [TEP-0076 Array results and indexing](https://github.com/tektoncd/community/pull/477)
+
+This feature would be considered alpha and would be
+[gated by the alpha flag](https://github.com/tektoncd/pipeline/blob/main/docs/install.md#alpha-features).
+
+### Notes/Caveats
+
+[See also TEP-0076 Array results and indexing](https://github.com/tektoncd/community/pull/477) for notes and caveats
+specific to supporting json results.
+
+* What if a Pipeline tries to use a object key that doesn't exist in a param or result?
+  * Some of this could be caught at creation time, i.e. pipeline params would declare the keys each object param will
+    contain, and we could validate at creation time that those keys exist
+  * For invalid uses that can only be caught at runtime, the PipelineRun would fail
+* What if a Task tries to write more keys to a result than the result declares it contains?
+  * The TaskRun would fail (alternatively we could decide to ignore the extra keys)
+* What if a Task tries to write less keys to a result than the result declares it contains?
+  * The TaskRun would fail
```

Comment:
Could we have the notion of "optional" keys ? does it make sense ? otherwise, agreeing less keys should probably be make it fail.

Reply:
hmm i tried to figure out how we'd express this with json schema and was surprised to see that by default all fields are optional unless they are explicitly `required` 🤔 

I'm a bit torn because i dont want to augment json schema, but at the same time i feel like defaulting to "required" is more consistent with our existing interfaces. we could allow them all to be optional by default but fail once variable replacement tries to use them? or we could imply a `required` section that includes all the values when its not present but that feels like a slippery slope 🤔 

seems like other ppl have run into similar concerns: https://stackoverflow.com/questions/16204871/what-is-the-difference-between-required-vs-optional-in-json-schema

Reply:
One thing that's nice about defaulting to optional is it makes it makes the default behavior safer for adding new fields. This will be particularly useful if we allow reusable schema definitions, since this would allow you to add a new field to a shared schema, then update the Tasks that use the field to now accept it.

e.g. if we wanted to add a new field to a Git resource type, optional fields would let phase the field in over time:
1. Add the field to the schema
2. Have producers / results start populating the field 
3. Consumers start using the field once it's populated.

> we could allow them all to be optional by default but fail once variable replacement tries to use them?
 
+1. It's a bit strange to have required optional fields, but I think this makes sense if we support shared schema types.

Reply:
ahhh that's a great point @wlynch , I hadn't thought of that